Improve callPackage error message!

infinisil opened the pull request:
###### Motivation for this change
Oftentimes arguments in package declarations are misspelled, leading to errors like
```
error: anonymous function at /home/infinisil/src/nixpkgs/pkgs/tools/text/ripgrep/default.nix:1:1
  called without required argument 'fetchFromGithub', at /home/infinisil/src/nixpkgs/lib/customisation.nix:69:16
```

which doesn't give any indication as to what the correct attribute is. Well with this change it does! The error now is
```
error: Function in /home/infinisil/src/nixpkgs/pkgs/tools/text/ripgrep/default.nix 
  called without required argument "fetchFromGithub", 
  did you mean "fetchFromGitHub" or "fetchFromGitLab"?
```

Which imo, is pretty damn fancy!

The implementation searches through all attributes in `pkgs` for one with a [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance) of 2 or less, catching most misspellings. Had to jump through some hoops to make this fast though.

Ping @alyssais @capsensitive @worldofpeace @Synthetica9 @roberth @Profpatsch

Better alternative to changes like https://github.com/NixOS/nixpkgs/pull/79854

###### Things done

- [ ] Wrote tests
- [x] Checked how this influenced performance (it shouldn't influence it much, since in the working case not much extra is done). <- That's correct, see https://github.com/NixOS/nixpkgs/pull/79877#issuecomment-586723868

samueldr commented on lib/customisation.nix:
```diff
@@ -117,8 +117,41 @@ rec {
+        # Put strings with shorter distance first
+        (lib.sort (x: y: lib.strings.levenshtein x arg < lib.strings.levenshtein y arg))
+        # Only take the 5 first results
+        (lib.take 3)
```
5 or 3? or is there something more?

roberth commented on lib/customisation.nix:
```diff
@@ -117,8 +117,41 @@ rec {
+      errorForArg = arg: "Function ${prettyLocation}"
+        + "called without required argument \"${arg}\"${prettySuggestions (getSuggestions arg)}";
+      error = lib.concatMapStringsSep "\n" errorForArg missingArgs;
+    in if missingArgs == [] then makeOverridable f allArgs else throw error;
```
This looks like it won't work for invocations that only become valid *after* overriding. I hope I'm wrong.
Also this is demanding `missingArgs` on all `callPackage` invocations. What's the performance impact?

Infinisil replied:
I guess you mean something like this:
```nix
(lib.callPackageWith {} ({ foo }: foo) {}).override { foo = 10; }
```

However this also throws an error even before this change.

Yeah the performance needs to be investigated, I think it's going to be minor though

roberth commented:
As much as I want this feature, I doubt that it should be implemented in Nixpkgs rather than Nix itself.
When implemented in Nix itself, this functionality will be available in *all* attrset-function invocations, rather than just callPackage, without any overhead; neither for the interpreter hot path nor for the Nixpkgs reader.

roberth commented:
This is the lookup in Nix: https://github.com/NixOS/nix/blob/c3181e21e75d9e31254cd97c8e56858be7072d5f/src/libexpr/eval.cc#L1161

Infinisil commented:
@roberth Unless `callPackageWith` becomes a builtin, I'm pretty sure this actually can't be implemented in Nix itself. All such functions are defined like `{ stdenv }` (*without* ellipsis), meaning exactly those arguments need to be passed in, no more. To do this, function arguments are intersected with the automatic arguments (`pkgs`), and only passing the result of that to the function. At this point, Nix doesn't know anything about the automatic arguments anymore -> it's unable to look those up.

What could be implemented in Nix however is the same thing for attribute access like `pkgs.fetchFromGithub`

grahamc commented:
I wonder what it would take to implement this in Nix itself, to get this goodness everywhere!

Infinisil commented:
Yeah so I'm pretty sure now. This only works due to the fact that `callPackageWith` has knowledge of all available packages in addition to the function arguments. To implement this in Nix would mean to implement `callPackageWith` as a builtin. Since this function depends on many other things however, this would mean implementing pretty much all of `lib/customisation.nix` as builtins. That might not even be too unreasonable, so if anybody wants to do this feel free.

Infinisil commented:
After some more tinkering I was able to come up with this change:
```diff
diff --git a/lib/customisation.nix b/lib/customisation.nix
index ac234e3b8c6..9944d3330f2 100644
--- a/lib/customisation.nix
+++ b/lib/customisation.nix
@@ -118,7 +118,10 @@ rec {
     let
       f = if lib.isFunction fn then fn else import fn;
       auto = builtins.intersectAttrs (lib.functionArgs f) autoArgs;
-    in makeOverridable f (auto // args);
+      allPotentialArgs = autoArgs // args;
+      filled = lib.mapAttrs (name: value: allPotentialArgs.${name})
+        (lib.filterAttrs (name: value: ! value) (lib.functionArgs f));
+    in makeOverridable f (filled // auto // args);
 
 
   /* Like callPackage, but for a function that returns an attribute
```

This results in this error now:
```
error: attribute 'fetchFromGithub' missing, at /home/infinisil/src/nixpkgs/lib/customisation.nix:122:43
(use '--show-trace' to show detailed location information)
```

So this change makes it look up all arguments in the available packages, resulting in an attribute access error instead of a function call error for missing arguments. With this, implementing such a feature for attribute access in Nix directly (as in `pkgs.fetchFromGithub` would throw a nice error), would automatically get it for `callPackagesWith` too! Unfortunately the error message doesn't show the location of the function anymore, which is unfortunate.

So, this could be implemented in Nix directly, either by:
- Implementing this error for attribute access and using above patch, only downside is the lost error location
- Implementing this error through a builtin `callPackageWith`, with the downside of being very complex and having no real benefit

If losing error location (without `--show-trace`) is acceptable, then using the former is most reasonable. If it is not, then this PR is the next best alternative, at no significant cost.

roberth commented:
`autoArgs` is huge, so you probably want to perform "two" lookups instead of doing a big merge:

    (name: value: autoArgs.${name} or args.${name} or autoArgs.${name})

The function location can be recovered by allowing functions as the first argument to `builtins.addErrorContext`. This does require `--show-trace`.

Alternatively, `builtins.throw`/`abort` can be changed to take a list which is concatMapped internally with a toString that does print function locations. This does require suggestions to be implemented in Nix, but if it improves performance for the hot path that's worth it.

The readability concern is minor if the big let binding is moved out into a separate function.

A possible benefit of generating suggestions in Nix is you can more easily add domain-specific logic. For example giving special treatment to common errors for which Levenshtein distance doesn't work too well, like adding/removing a `lib` prefix. However, I think there's really not that much preventing Nix itself from doing that too.

(If we go with a Nix implementation) Did you try canonicalizing the names before running Levenshtein? You could
 - lowercase everything
 - remove `_` and `-`
 - remove `lib` prefix
 - remove version number suffixes (maybe too much loss, maybe slow?)

Infinisil commented:
@roberth I like your idea of being able to have more domain-specific behavior with an implementation in Nix code. I think doing those suggestions here would decrease code readability a bit too much for now, and the current approach has some leeway for things like `python-28 -> python38/python27` still.

So, I'm kind of convinced this PR is a good idea because:
- Implementing this builtin into Nix with above patch would lose error location without workarounds
- Implementing this builtin into Nix with a `callPackageWith` would be very complex
- Doing this in Nix code directly is relatively simple and doesn't lead to worse errors
- Doing this in Nix code directly allows for more flexibility and domain-specifics in the future
- This PR has real benefits without any costs. Just because it doesn't implement these suggestions for all of Nix doesn't make those benefits any worse

Imo this can be merged after performance is confirmed to not be dramatically worse (which I doubt)

Infinisil commented:
The time to evaluate `nix-instantiate nixos/release-combined.nix -A tested` with 5 samples (stats calculated with [sta](https://github.com/NixOS/nixpkgs/pull/78141)):

| state | min | max | mean | stddev | stderr |
| --- | --- | --- | --- | --- | --- |
| Before PR | 211.659 | 218.022 | 215.039 | 2.77587 | 1.24141 |
| After PR | 211.653 | 217.456 | 214.695 | 2.00313 | 0.895827 |

So I think it's safe to say that performance is influenced minimally by this.

grahamc commented:
I like this PR.

Given the recent evaluation memory wall we hit (and are still having trouble with) I am -1 on this PR merging _right now_. I would like to see that stabilize before adding this. I know it may not make it worse, but I don't want to tempt fate today.

edolstra commented:
It's a nice feature, but Nix is not a general purpose language so this definitely should not be implemented in Nix code.

roberth commented:
@Infinisil The evaluation time of your benchmark may be dominated by time spent in the module system. Benchmarking Nixpkgs by itself should help to make the results more significant, considering that standard error estimate is currently rather high.

Infinisil commented:
@edolstra I guess you're referring to the levenshtein distance being implemented in nix code? That's the only thing that's "general-purpose" here. Yeah I guess a `builtins.levenshtein[AtMost]` might be a good idea in Nix itself, but there is no harm having this implemented in Nix code.

edolstra commented:
@Infinisil Yes, exactly. Nix isn't really suited for implementing anything that has the word "algorithm" in it :-) E.g. Nix doesn't have a type system, the evaluator is not very fast, etc. 

I don't think the solution is to have ` builtins.levenshtein`. Rather, the evaluator could give better error "undefined variable" messages by using Levenshtein internally.

Infinisil commented:
@edolstra As explained above already, this either has the disadvantage of: Losing error context if using https://github.com/NixOS/nixpkgs/pull/79877#issuecomment-584978908 or significant complexity in Nix builtins using https://github.com/NixOS/nixpkgs/pull/79877#issuecomment-584949864. If you know of a way to get an error message without losing context and without too much complexity, let me know, because I don't see it.

And as such, I still think this PR is the best way to implement this.

roberth commented:
Isn't `callPackage` a thing we can get rid of?
This works fine:
```
> import ./foo.nix pkgs      
error: anonymous function at /home/user/h/nixpkgs/foo.nix:1:1 called without required argument 'qux', at (string):1:1
```

Is there any reason to use reflection and `intersectAttrs`, other than
 - having a bunch of function definitions without `{ , ... }`
 - possibly improved performance when `callPackage x { someOverride = y; }` is used

Both of these seem solvable.
Most `callPackage` invocations don't seem to have overrides, so shallow copying `pkgs` only a few times shouldn't be a problem and it does save many `intersectAttrs` calls + allocations.

What we get in returns is better error messages as implemented in Nix. We want to have it there for the general case of attrset invocations anyway. Furthermore, we simplify Nixpkgs by getting rid of needless use of reflection and we get the same behavior as with the module system, which also requires `...`.

Also we can replace this entire nix pill https://nixos.org/nixos/nix-pills/callpackage-design-pattern.html by a footnote about how laziness is working really well. No need to waste people's time on legacy.

Please correct me if I'm wrong :)

Infinisil commented:
@roberth For one, `callPackage` also calls `makeOverridable` which adds `.override` and `.overrideAttrs`, so a bit more would still be needed.

I like the idea of redesigning how packages are declared in nixpkgs, the current way seems clumsy in many ways (multiple dependency listings, what's up with optional dependencies, what's up with configuration, why `callPackage` with overrides, why not auto-call all these package files). However anything like that should have some serious thought put into it before it is applied, because it involves tree-wide changes and requires people to know about the new ways. As you suggest, we might get this feature for free once that is done and Nix implements such suggestions, but this will be a long ways away.

roberth commented:
@Infinisil I agree, except for the long ways away. It depends on the migration path we choose, but either way it's going to be a 'mechanical' process; simple and boring, which is good.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/fetchfromgithub-in-configuration-nix/5927/10

grahamc commented:
I think a few people have misunderstood @edolstra's position here. He doesn't want to implement this language in code evaluated by the evaluator -- ie: he doesn't want this solution to be done with the Nix Expression Language. Instead, he would like to see the solution implemented as a patch to the Nix Evaluator's error messages, in the NixOS/nix repo.

edolstra commented:
If we're on the topic of simplifying package expressions, we should ask whether packages should be functions at all. The main reason why they are functions is because you don't want to define all packages in one giant file. But things become much simpler if you *do* put everything in a giant file, see for example `perl-packages.nix` which doesn't use package functions (dependencies aren't passed around through function arguments but are obtained from the lexical environment). Since such a file becomes unwieldy, we could an [include mechanism](https://github.com/edolstra/nix/commit/571430e1b42010114780ec4c309da3193bdc345a), i.e.
```
firefox = include ./path/to/firefox.nix;
libfoo = include ./path/to/libfoo.nix;
libbar = include ./path/to/libbar.nix;
stdenv = ...;
```
where `firefox.nix` looks like
```
stdenv.mkDerivation {
  buildInputs = [ libfoo libbar ... ];
}
```
i.e. it doesn't take any arguments but just expects to be included in a lexical environment containing `stdenv`, `libfoo`, etc.

This style does break `.override` but arguably that's a good thing.

Another possibility is Bazel-style dependencies where packages themselves specify where to get their dependencies, so something like
```
buildInputs =
  [ (import ../path/to/libfoo.nix)
    (import ../path/to/libbar.nix)
    ...
  ];
```
but this changes semantics a lot.

7c6f434c commented:
>This style does break `.override` but arguably that's a good thing.

We even use overrides in Nixpkgs proper, and this _is_ useful functionality.

roberth commented:
Re `include`: I'm not sure if I want to get used to the idea of such specific scoping rules. Regardless, as a feature it doesn't seem to interact with improving lookup error messages in Nix.

To take advantage of the (native) improved error messages, some refactoring is needed. This refactoring can be gradual and does not need to remove overriding. Radical changes like `include` and removing overriding can be discussed separately. Let's focus this discussion on the steps we can take to make the error message improvement happen.

From what I gather, the following needs to be done:
 - add a Levenshtein lookup function to Nix
 - use it in the error messages for attrset and potentially scope lookups
 - start avoiding `intersectAttrs` before attrset application

This doesn't necessarily mean I'm opposed to the more radical proposals. I'd just like us to make progress rather than tangle things up.

@Infinisil would you like to contribute this feature to `nix`?

Infinisil commented:
Not really, C++ is not my turf. And I still think losing error context (without `--show-trace`) from this refactoring would be inferior to what is done here, but maybe somebody can find a way around that.

roberth commented:
Source location doesn't seem to be a problem anymore.
```
nix-repl> simplisticCallPackage = file: makeOverridable (import file) pkgs     

nix-repl> simplisticCallPackage ./foo.nix
error: anonymous function at /home/user/h/nixpkgs/foo.nix:1:1 called without required argument 'qux', at /home/user/h/nixpkgs/lib/customisation.nix:69:16
```

which is at least as good as the current

```
nix-repl> callPackage ./foo.nix {}
error: anonymous function at /home/user/h/nixpkgs/foo.nix:1:1 called without required argument 'qux', at /home/user/h/nixpkgs/lib/customisation.nix:69:16
```

Prior has all information necessary for Nix to give an improved error message, because `makeOverridable` passes all arguments. It does not use `intersectAttrs` and as shown in the output, which is without `--show-trace`, source location is available. Did I miss something?

Infinisil commented:
@roberth It requires `foo.nix` to have `...` in the argument list, which has been discouraged for as long as I can remember

roberth commented:
@Infinisil I'd guess it's because the other arguments were not available, due to `intersectAttrs`. With the new function those will be available. Does anyone know the reason(s) behind this discouragement?

Infinisil commented:
Ultimately I think it would be possible to have a better implementation of this, but for now I think this is the best we can reasonably do as a small step towards better such error messages. This change is rather forgiving too, merging it doesn't cost us anything really, and we can also revert it later without any repercussions, potentially replacing it with something better. This change seems worthwhile, but not perfect, to me.

For now I won't be doing a re-architecting of nixpkgs for this, nor will I be writing an C++ implementation with above-mentioned problems.

alyssais commented:
What's the path forward here, then?  Do we merge or close?

Infinisil commented:
I'm obviously biased, but I'd say merge, because I don't see any downside to this. (Though if we decide to merge, I'll write some tests and stuff first)

Profpatsch commented:
I’m also +1 because I think @Infinisil has demonstrated that the change doesn’t hurt evaluation times while improving usability.

Instead of dreaming about how things should be, we should improve them now in the knowledge that we can remove the code in the future when nix itself is ready.

Infinisil commented:
I discovered that actually in one rare case this decreases error clarity. Namely when you do a `callPackage` with a function directly, not with a path, e.g.
```nix
callPackage ({ foo }: {}) {}
```

Previously this threw the error
```
error: anonymous function at /home/infinisil/src/nixpkgs/cp.nix:4:21 called without required argument 'foo', at /home/infinisil/src/nixpkgs/lib/customisation.nix:69:16
```

whereas currently with this PR it's just 
```
error: Function called without required argument "foo"
```

I have now submitted a PR to Nix that will allow this error to be improved with location info: https://github.com/NixOS/nix/pull/3468, after which the error can look like
```
error: Function called without required argument "foo" at /home/infinisil/src/nixpkgs/cp.nix:4
```

In addition, that PR will allow line-specific error messages even when files are passed (which wouldn't be possible otherwise).

Once that PR is merged, I'll extend this one with a patch to support it.

infinisil committed:
lib/strings: Add levenshtein distance functions

Adds some functions related to string similarity:
- lib.strings.commonPrefixLength
- lib.strings.commonSuffixLength
- lib.strings.levenshtein
- lib.strings.levenshteinAtMost

infinisil committed:
lib/tests: Add tests for levenshtein functions

infinisil committed:
lib/customization: Improve callPackage error message for missing args

This uses the levenshtein distance to look through all possible
arguments to find ones that are close to what was requested:

  error: Function in /home/infinisil/src/nixpkgs/pkgs/tools/text/ripgrep/default.nix
    called without required argument "fetchFromGithub",
    did you mean "fetchFromGitHub" or "fetchFromGitLab"?

With https://github.com/NixOS/nix/pull/3468 (in current nixUnstable) the error
message becomes even better, adding line location info

Infinisil commented:
I am conflicted. With the recent error message enhancements in https://github.com/NixOS/nix/pull/3773, the error without this change now looks like
```
error: --- TypeError ------------------------------------------------------------------------------------------------------- nix-instantiate
at: (69:16) in file: /home/infinisil/src/nixpkgs/lib/customisation.nix

    68|     let
    69|       result = f origArgs;
      |                ^
    70|

anonymous function at /home/infinisil/src/nixpkgs/pkgs/tools/text/ripgrep/default.nix:1:1
  called without required argument 'fetchFromGitHud'
```

Which while pretty, doesn't actually help very much, and I doubt the error message proposal can help with that without major complexity as described in https://github.com/NixOS/nixpkgs/pull/79877#issuecomment-586766416 and co.

With this PR, the error looks like
```
error: --- ThrownError ----------------------------------------------------------------------------------------------------- nix-instantiate
Function called without required argument "fetchFromGitHud"
  at /home/infinisil/src/nixpkgs/pkgs/tools/text/ripgrep/default.nix:2,
  did you mean "fetchFromGitHub" or "fetchFromGithub"?
```

Which is not as pretty, but actually helps more with the error. So I think it would still be a good idea to merge this, because other than the error messages in unstable being prettier, nothing changed.

Thoughts?

Profpatsch commented:
@domenkozar maybe we can build that directly into the new error messages? It’s a simple distance function to compare with symbols in scope.

Infinisil commented:
@Profpatsch It's unfortunately not that simple, because in the current code, no Nix builtin that would know the scope of supported arguments is called. I think it might be possible if we introduce a `builtins.callFunctionIntersectingArgs f pkgs` which would behave like `f (builtins.intersectAttrs (builtins.functionArgs f) pkgs)`, but could support a better error message.

Profpatsch commented:
builtins are relatively easy to add, so that might be a good idea. cc
@edolstra

On Tue, Jul 28, 2020 at 11:04 PM Silvan Mosberger <notifications@github.com>
wrote:

> @Profpatsch <https://github.com/Profpatsch> It's unfortunately not that
> simple, because in the current code, no Nix builtin that would know the
> scope of supported arguments is called. I think it might be possible if we
> introduce a builtins.callFunctionIntersectingArgs f pkgs which would
> behave like f (builtins.intersectAttrs (builtins.functionArgs f) pkgs),
> but could support a better error message.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/79877#issuecomment-665281821>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYB5ZXXWEW245NLZPEFQVLR544N7ANCNFSM4KTNQRHQ>
> .
>

Infinisil commented:
I have started implementing a Nix builtin that would allow this. Work is in this branch: https://github.com/Infinisil/nix/tree/callFunctionDefaultArgs

Profpatsch commented:
> I have started implementing a Nix builtin that would allow this. Work is in this branch: https://github.com/Infinisil/nix/tree/callFunctionDefaultArgs

Did you get it to a point where it’s mergable? I say we should still go ahead with this patch and then do it like with other lib functions when a builtin is introduced, conditionally use the builtin if it exists.

Infinisil commented:
@Profpatsch Except for missing tests, I think it would be mergeable. However the functionality that suggests alternative attributes isn't implemented yet, only a builtin that would support this.

Profpatsch commented:
I propose we merge then.

roberth commented:
I don't think it's a particularly good solution, but it solves a problem. If it causes any issues, it can be reverted.

For readability, you could factor the error checking and throw into a helper function as much as possible. That should leave a concise implementation of what `callPackageWith` really does. Should be easy because you need to rebase anyway.
If you could do that, you have my approval.

edolstra commented:
I still don't think that the Nixpkgs standard library is the right place for functions like `levenshtein`. This is complexity that we need to support forever once added.

jtojnar commented:
Could we name it `_internalLevenshtein`, or require the user to call it as `_internalLevenshtein {__acceptThisFunctionIsInternalAndCanBeRemovedAtAnyTime = true; } a b` to rid us of such support requirements?

Profpatsch commented:
Just move it to a let binding inside the place it’s used instead of
exposing it from lib?

On Fri, Oct 9, 2020 at 12:10 PM Jan Tojnar <notifications@github.com> wrote:

> Could we name it _internalLevenshtein, or require the user to call it as _internalLevenshtein
> {__acceptThisFunctionIsInternalAndCanBeRemovedAtAnyTime = true; } a b to
> rid us of such support requirements?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/79877#issuecomment-706094580>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYB5ZUBXBLCABR2NRA6FFDSJ3ORRANCNFSM4KTNQRHQ>
> .
>

jtojnar commented:
Private definitions are harder to run tests for.

Infinisil commented:
But also, I don't think there's a problem with having a generic `levenshtein` function or the likes in nixpkgs. These functions don't need constant maintenance. And even if they do, it's probably much less than all other parts of nixpkgs, and I'd be happy to do it.

Infinisil commented:
I wouldn't mind this being merged, though I might continue my attempt at implementing this in Nix directly soon, which then would make this PR redundant.

Profpatsch commented:
So let’s get rid of the generic levenshtein function in this PR and only
add the specialized version for now, it seems like this was the only
objection left.

On Sat, Oct 10, 2020 at 12:17 AM Silvan Mosberger <notifications@github.com>
wrote:

> I wouldn't mind this being merged, though I might continue my attempt at
> implementing this in Nix directly soon, which then would make this PR
> redundant.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/79877#issuecomment-706420403>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYB5ZQVTCVKXVDZALGQGILSJ6DV5ANCNFSM4KTNQRHQ>
> .
>

Profpatsch commented:
It still works for me:

```
> nix-build -A ripgrep
error: Function called without required argument "fetchFromGitHu" at /home/philip/nixpkgs/pkgs/tools/text/ripgrep/default.nix:2, did you mean "fetchFromGitHub", "fetchFromGitea" or "fetchFromGithub"?
```

I would merge within a few days.
